### DISCUSSION TITLE
Connections and sessions managment in HA mode

### DISCUSSION THREAD

**Discussion by VladOrlov** [apache/kyuubi#5124] [Q&A] [answered]
Kyuubi's doc HA contains this information:
With [Kyuubi Hive JDBC Driver](https://mvnrepository.com/artifact/org.apache.kyuubi/kyuubi-hive-jdbc) or vanilla Hive JDBC Driver, a client can specify service discovery mode in JDBC connection string, i.e. serviceDiscoveryMode=zooKeeper; and set zooKeeperNamespace=kyuubi;, then it can randomly pick one of the Kyuubi service uris from the specified ZooKeeper addresses in the /kyuubi path.

So each new connection from one user will be randomly redirected to available Kyuubi service and new session will be created?

Also we have connections issue in HA mode on Kubernetes. Zookeeeper contains pods' internal IP so clieant can't reach Kyuubi. Could you please explaine how correctly setup Kyuubi on Kubernetes in HA mode?

**Comment by pan3793** [comment]
> So each new connection from one user will be randomly redirected to available Kyuubi service and new session will be created?

Yes.

> Zookeeeper contains pods' internal IP so clieant can't reach Kyuubi.

Maybe https://github.com/apache/kyuubi/pull/5015 can answer your question.

Further, you are not enforced to use ZK for Kyuubi Server HA, Ingress/LoadBalance are fine too. (Note that, ZK is still required for engine discovery, in this case, it's only used for Kyuubi internally, assume your Kyuubi server and Spark runs on the same K8s cluster, should not have network reachable issues.)

**Comment by hanna-liashchuk** [comment]
In this case, share level USER is "half" working? Since you have to be on the same machine to share the engine and zookeeper not necesarily will redirect you to the same server. Meaning - having 3 servers and 3 connects with share level user can result in 3 separate engines, right?

**Reply by pan3793** [reply]
No. An external zookeeper cluster is required for the engine to register itself, so that all Kyuubi servers could find it.

**Reply by hanna-liashchuk** [reply]
but what if kyuubi servers are in k8s and spark is in k8s. in this case an engine is registered in the same pod as server. 
If there is a zookeeper (not embeded) and user connects with serviceDiscoveryMode=zooKeeper and share level USER, it will always go to the same server?

**Reply by pan3793** [reply]
> in this case an engine is registered in the same pod as server.

It's only true in spark client mode. But it does not matter, even though the engine(Spark driver) runs in server1's pod, it still could be connected from server2 as long as the server2 can find the engine's connection info from zk, 

> it will always go to the same server

No, randomly pick one. But all Kyuubi servers could see the registered engine in ZK.

**Reply by hanna-liashchuk** [reply]
> 
> It's only true in spark client mode. 

Is there another option for an interactive session?

Soo, let me wrap it up. A user will fo to Zookeeper, Zookeeper will send the user to a random Kyuubi server, the server will see that the user has share level USER and it will know the address of already available engine and send the user to that server? 
i.g. Client -> Zookeeper -> Kyuubi Server 0 -> Kyuubi Server 1

**Reply by pan3793** [reply, marked as answer]
> Is there another option for an interactive session?

Kyuubi supports all modes that Spark supported. including local, client, cluster.

> Client -> Zookeeper -> Kyuubi Server 0 -> Kyuubi Server 1

No. Zookeeper is not only used for Server register, but also engine register, the key point here is the routing/communication mechanism between Kyuubi Server and Kyuubi Engine(Spark Driver)

<img width="600" alt="image" src="https://github.com/apache/kyuubi/assets/26535726/25abb1f1-84a8-4842-a668-4157e873cfa1">

<img width="757" alt="image" src="https://github.com/apache/kyuubi/assets/26535726/be6c1fee-1e8b-4215-8647-81d72e528b1a">
